frontend/catalyst/jax_tracer.py

```diff
@@ -106,6 +113,59 @@ def _eval_jaxpr(*args):
 }
 
 
+def _promote_jaxpr_types(types: List[List[ShapedArray]]) -> List[ShapedArray]:
+    # TODO: We seem to use AbstractQreg incorrectly, so JAX doesn't recognize it as a valid abstact
+    # value. One need to investigate how to use it correctly and remove the condition [1].
+    assert len(types) > 0, "Expected one or more set of types"
+    assert all(len(t) == len(types[0]) for t in types), "Expected matching number of arguments"
+    with_qregs = all(isinstance(t[-1], AbstractQreg) for t in types)
+    if with_qregs:  # [1]
+        types = [t[:-1] for t in types]
+    results = list(map(partial(reduce, promote_types), zip(*types)))
+    return results + ([AbstractQreg()] if with_qregs else [])
+
+
+def _apply_result_type_conversion(
+    jaxpr: ClosedJaxpr, target_types: List[ShapedArray]
+) -> ClosedJaxpr:
+    with_qreg = isinstance(target_types[-1], AbstractQreg)
+    with EvaluationContext(EvaluationMode.CLASSICAL_COMPILATION) as ctx:
+        with EvaluationContext.frame_tracing_context(ctx) as trace:
+            in_tracers = _input_type_to_tracers(trace.new_arg, jaxpr.in_avals)
+            out_tracers = [
+                trace.full_raise(t) for t in eval_jaxpr(jaxpr.jaxpr, jaxpr.consts, *in_tracers)
+            ]
+            out_tracers_, target_types_ = (
+                (out_tracers[:-1], target_types[:-1]) if with_qreg else (out_tracers, target_types)
+            )
+            out_promoted_tracers = [
+                (convert_element_type(tr, ty) if _abstractify(tr).dtype != ty else tr)
+                for tr, ty in zip(out_tracers_, target_types_)
+            ]
+            jaxpr2, _, consts = ctx.frames[trace].to_jaxpr2(
+                out_promoted_tracers + ([out_tracers[-1]] if with_qreg else [])
+            )
+    return ClosedJaxpr(jaxpr2, consts)
+
+
+def unify_result_types(jaxprs: List[ClosedJaxpr]) -> List[ClosedJaxpr]:
+    """Unify result types of the jaxpr equations given.
+    Args:
+        jaxprs (list of ClosedJaxpr): Source JAXPR expressions. The expression results must have
+                                      matching pytree-shapes and numpy array shapes but dtypes might
+                                      be different.
```

Comment:
Optional: Should we assert this?

Reply:
We do assertions in the `_promote_jaxpr_types`

Reply:
Don't these assertions:
```py
assert len(types) > 0, "Expected one or more set of types"
assert all(len(t) == len(types[0]) for t in types), "Expected matching number of arguments"
```
where `types = [j.out_avals for j in jaxprs]`,

only check that the number of results from each jaxpr is the same? That is it doesn't check whether the pytrees match, nor whether array shapes match, which the docstring states as a requirement.

Reply:
Ah, correct! I will add those.

Reply:
I have added the shape assertion, but the pytree one is not that easy. We don't need the Pytrees when we do quantum tracing. Adding them just to check for equivalence here doesn't look feasible. We do check it [elsewhere](https://github.com/PennyLaneAI/catalyst/blob/96db366800d5ab2ff67024c27fcfaa26ed2d66d8/frontend/catalyst/pennylane_extensions.py#L1171) though.

Reply:
But that means we don't need to mention pytrees in the comment so I have updated it.

Reply:
I don't think it's important that the Pytrees match, otherwise we wouldn't know which Pytree to restore as the result of the conditional operation.

Reply:
Oh sorry, I misspelled. I meant "I _do_ think it's important that the PyTrees match". I think we should add those test cases @grwlf

Reply:
Addressed here https://github.com/PennyLaneAI/catalyst/pull/351